File: src/main/java/em/Lexer.java
package em;

import java.io.*;
import java.util.ArrayList;

/**
 * Created by andrew on 9/18/16.
 */
public class Lexer {

    private FileInput input;
    private ArrayList<Lexeme> lexemes;
    private File file;

    public Lexer(String filename){
        this(new File(filename));
    }

    public Lexer(File file){
        this.file = file;
        try{
            input = new FileInput(file);
        } catch(IOException e){
            Logger.error("Error in reading file");
            e.printStackTrace();
            System.exit(1);
        }

        //Create the list for the lexemes
        lexemes = new ArrayList<>();
    }

    public ArrayList<Lexeme> getLexemes(){
        boolean go;
        do{
            go = lex();
        }while(go);
        return lexemes;
    }

    public boolean lex(){
        input.skipWhitespace();

        // Read a codepoint in
        int row = input.getRow();
        int col = input.getCol();
        int in = input.read();

        // end of file, stop lexing
        if(in <= 0 || in == 65535){
            return false;
        }

        if(in < Character.MAX_VALUE){
            switch((char) in){
                case '.':
                    lexemes.add(new Lexeme(LexemeType.DOT, row, col, this.file));
                    return true;
                case ';':
                    lexemes.add(new Lexeme(LexemeType.SEMICOLON, row, col, this.file));
                    return true;
                case ',':
                    lexemes.add(new Lexeme(LexemeType.COMMA, row, col, this.file));
                    return true;
                case '(':
                    lexemes.add(new Lexeme(LexemeType.OPEN_PAREN, row, col, this.file));
                    return true;
                case ')':
                    lexemes.add(new Lexeme(LexemeType.CLOSE_PAREN, row, col, this.file));
                    return true;
                case '{':
                    lexemes.add(new Lexeme(LexemeType.OPEN_CURLY, row, col, this.file));
                    return true;
                case '}':
                    lexemes.add(new Lexeme(LexemeType.CLOSE_CURLY, row, col, this.file));
                    return true;
                case '[':
                    lexemes.add(new Lexeme(LexemeType.OPEN_BRACKET, row, col, this.file));
                    return true;
                case ']':
                    lexemes.add(new Lexeme(LexemeType.CLOSE_BRACKET, row, col, this.file));
                    return true;
                case '*':
                    lexemes.add(new Lexeme(LexemeType.MULTIPLY, row, col, this.file));
                    return true;
                case '%':
                    lexemes.add(new Lexeme(LexemeType.MODULO, row, col, this.file));
                    return true;
                default:
                    input.pushback(in);
                    if(lexMultiCharacterOperators()){
                        return true;
                    }
                    if(Character.isDigit(in)){
                        return lexNumber();
                    }
                    if(in == '"'){
                        return lexString();
                    }
            }
        }
        //Check for identifiers and keywords
        if(in >= Character.MAX_VALUE){
            input.pushback(in);
        }

        return lexKeywordsAndIdentifiers();
    }

    /*Specialized lexing functions*/

    /**
     * Parse all the non-single character operators
     * @return Whether a lexeme was parsed or not
     */
    private boolean lexMultiCharacterOperators(){
        int row = input.getRow();
        int col = input.getCol();
        int in = input.read();
        if(in == '+'){
            int test = input.read();
            if(test == '+'){
                lexemes.add(new Lexeme(LexemeType.INCREMENT, row, col, this.file));
            }
            else{
                input.pushback(test);
                lexemes.add(new Lexeme(LexemeType.PLUS, row, col, this.file));
            }
            return true;
        }
        if(in == '-'){
            int test = input.read();
            if(test == '-'){
                lexemes.add(new Lexeme(LexemeType.DECREMENT, row, col, this.file));
            }
            else if(Character.isDigit(test)){
                //For handling negative numbers
                input.pushback(test);
                input.pushback(in);
                return lexNumber();
            }
            else if(test == '>'){
                lexemes.add(new Lexeme(LexemeType.ARROW, row, col, this.file));
            }
            else{
                input.pushback(test);
                lexemes.add(new Lexeme(LexemeType.MINUS, row, col, this.file));
            }
            return true;
        }
        if(in == '/'){
            int test = input.read();
            if(test == '/'){
                input.skipLine();
            }
            else{
                input.pushback(test);
                lexemes.add(new Lexeme(LexemeType.DIVIDE, row, col, this.file));
            }
            return true;
        }
        if(in == '>'){
            int test = input.read();
            if(test == '='){
                lexemes.add(new Lexeme(LexemeType.GTE, row, col, this.file));
            }
            else{
                input.pushback(test);
                lexemes.add(new Lexeme(LexemeType.GT, row, col, this.file));
            }
            return true;
        }
        if(in == '<'){
            int test = input.read();
            if(test == '='){
                lexemes.add(new Lexeme(LexemeType.LTE, row, col, this.file));
            }
            else{
                input.pushback(test);
                lexemes.add(new Lexeme(LexemeType.LT, row, col, this.file));
            }
            return true;
        }
        if(in == '&'){
            int test = input.read();
            if(test == '&'){
                lexemes.add(new Lexeme(LexemeType.AND, row, col, this.file));
                return true;
            }
            return false;
        }
        if(in == '|'){
            int test = input.read();
            if(test == '|'){
                lexemes.add(new Lexeme(LexemeType.OR, row, col, this.file));
                return true;
            }
            return false;
        }
        if(in == '='){
            int test = input.read();
            if(test == '='){
                lexemes.add(new Lexeme(LexemeType.EQUALITY, row, col, this.file));
            }
            else{
                input.pushback(test);
                lexemes.add(new Lexeme(LexemeType.ASSIGN, row, col, this.file));
            }
            return true;
        }
        if(in == '!'){
            int test = input.read();
            if(test == '='){
                lexemes.add(new Lexeme(LexemeType.NOT_EQUALS, row, col, this.file));
            }
            else{
                input.pushback(test);
                lexemes.add(new Lexeme(LexemeType.NOT, row, col, this.file));
            }
            return true;
        }
        input.pushback(in);

        return false;
    }


    /**
     * Parse lexemes for language keywords and identifiers
     * @return Where a keyword or identifier was parsed
     */
    private boolean lexKeywordsAndIdentifiers(){
        int row = input.getRow();
        int col = input.getCol();
        int in = input.read();
        StringBuilder sb = new StringBuilder();

        while(Character.isDigit(in)
                || Character.isAlphabetic(in)
                || Character.isSupplementaryCodePoint(in)
                || in == '_'
                || in == '-'){
            sb.appendCodePoint(in);
            in = input.read();
        }
        //Push back whatever broke the loop
        input.pushback(in);

        //Check all the keywords
        String inputString = sb.toString();
        if(inputString.equals("var")){
            lexemes.add(new Lexeme(LexemeType.VAR, row, col, this.file));
            return true;
        }
        else if(inputString.equals("function")){
            lexemes.add(new Lexeme(LexemeType.FUNCTION, row, col, this.file));
            return true;
        }
        else if(inputString.equals("if")){
            lexemes.add(new Lexeme(LexemeType.IF, row, col, this.file));
            return true;
        }
        else if(inputString.equals("else")){
            lexemes.add(new Lexeme(LexemeType.ELSE, row, col, this.file));
            return true;
        }
        else if(inputString.equals("for")){
            lexemes.add(new Lexeme(LexemeType.FOR, row, col, this.file));
            return true;
        }
        else if(inputString.equals("while")){
            lexemes.add(new Lexeme(LexemeType.WHILE, row, col, this.file));
            return true;
        }
        else if(inputString.equals("do")) {
            lexemes.add(new Lexeme(LexemeType.DO, row, col, this.file));
            return true;
        }
        else if(inputString.equals("true")) {
            lexemes.add(new Lexeme(LexemeType.TRUE, row, col, this.file));
            return true;
        }
        else if(inputString.equals("false")) {
            lexemes.add(new Lexeme(LexemeType.FALSE, row, col, this.file));
            return true;
        }
        else if(inputString.equals("null")) {
            lexemes.add(new Lexeme(LexemeType.NULL, row, col, this.file));
            return true;
        }
        else if(inputString.equals("lambda")) {
            lexemes.add(new Lexeme(LexemeType.LAMBDA, row, col, this.file));
            return true;
        }
        else if(inputString.equals("class")) {
            lexemes.add(new Lexeme(LexemeType.CLASS, row, col, this.file));
            return true;
        }
        else if(inputString.equals("is")) {
            lexemes.add(new Lexeme(LexemeType.IS, row, col, this.file));
            return true;
        }
        else{
            //Must be an identifier, if it didn't match any keyword
            lexemes.add(new Lexeme(LexemeType.IDENTIFIER, inputString, row, col, this.file));
            return true;
        }
    }

    /**
     * Lex a string literal
     * @return If a String lexeme was parsed
     */
    private boolean lexString(){
        int row = input.getRow();
        int col = input.getCol();
        Logger.silly("Lexing a string");
        int in = input.read();
        if(in == '"'){
            StringBuilder sb = new StringBuilder();
            in = input.read();
            do{
                if(in != '\\'){
                    sb.appendCodePoint(in);
                }
                in = input.read();
                if(in == '\\'){
                    int test = input.read();
                    sb.appendCodePoint(test);
                }
            }while(in != '"' && in != -1);
            lexemes.add(new Lexeme(LexemeType.STRING, sb.toString(), row, col, this.file));
            return true;
        }
        else{
            input.pushback(in);
            return false;
        }
    }

    /**
     * Lex a number
     * @return If a number was parsed
     */
    private boolean lexNumber(){
        int row = input.getRow();
        int col = input.getCol();
        Logger.silly("Lexing a number");
        StringBuilder sb = new StringBuilder();
        int in = input.read(); //This'll either be a "-" or a digit
        do{
            sb.appendCodePoint(in);
            in = input.read();
        }while(Character.isDigit(in));

        //Push back the non-digit character
        input.pushback(in);
        lexemes.add(new Lexeme(LexemeType.INTEGER, Integer.parseInt(sb.toString()), row, col, this.file));
        return true;
    }
}

File: src/lexer.rs
use lexeme::Lexeme;
use lexeme_type::LexemeType;
use pushback_reader::PushbackCharReader;
use std::cell::RefCell;
use std::io::Read;
use std::io::Seek;
use std::rc::Rc;

pub struct Lexer<S>{
	source: Rc<RefCell<S>>,
	reader: PushbackCharReader<S>
}


impl <S: Read+Seek+Clone> Lexer<S>{
	pub fn new(src: S) -> Lexer<S>{
		return Lexer{
			reader: PushbackCharReader::new(src.clone()),
			source: Rc::new(RefCell::new(src))
		}
	}

	pub fn lex(&mut self) -> Option<Lexeme<S>>{
		self.reader.skip_whitespace();

		//Get our character ready
		let row = self.reader.get_line();
		let col = self.reader.get_col();
		let character = self.reader.read();
		
		let s_clone = self.source.clone();

		if let Some(c) = character{
			match c {
				'.' => Some(Lexeme::new(s_clone, LexemeType::DOT, row, col)),
				';' => Some(Lexeme::new(s_clone, LexemeType::SEMICOLON, row, col)),
				',' => Some(Lexeme::new(s_clone, LexemeType::COMMA, row, col)),
				'(' => Some(Lexeme::new(s_clone, LexemeType::OPEN_PAREN, row, col)),
				')' => Some(Lexeme::new(s_clone, LexemeType::CLOSE_PAREN, row, col)),
				'{' => Some(Lexeme::new(s_clone, LexemeType::OPEN_CURLY, row, col)),
				'}' => Some(Lexeme::new(s_clone, LexemeType::CLOSE_CURLY, row, col)),
				'[' => Some(Lexeme::new(s_clone, LexemeType::OPEN_BRACKET, row, col)),
				']' => Some(Lexeme::new(s_clone, LexemeType::CLOSE_BRACKET, row, col)),
				'*' => Some(Lexeme::new(s_clone, LexemeType::MULTIPLY, row, col)),
				'%' => Some(Lexeme::new(s_clone, LexemeType::MODULO, row, col)),
				'"' => {self.reader.pushback(c); self.lex_string()},
				_ => {
					self.reader.pushback(c);
					self.lex_multichar_op().or_else( || {
						self.lex_number().or_else( || {
							self.lex_keywords_and_identifiers()
						})
					})
				}
			}
		}
		else{
			None
		}
	}

	fn lex_multichar_op(&mut self) -> Option<Lexeme<S>>{
		println!("Lexing multichar op");
		//Get our character ready
		let row = self.reader.get_line();
		let col = self.reader.get_col(); 
		let c = self.reader.read().unwrap();

		let s_clone = self.source.clone();

		match c {
			'+' => {
				let test = self.reader.read();
				match test {
					Some(t) if t =='+' => Some(Lexeme::new(s_clone, LexemeType::INCREMENT, row, col)),
					Some(t) => {
						self.reader.pushback(t); 
						Some(Lexeme::new(s_clone, LexemeType::PLUS, row, col))
					},
					None => None
				}
			},
			'-' => {
				let test = self.reader.read();
				match test {
					Some(t) if t == '-' => Some(Lexeme::new(s_clone, LexemeType::DECREMENT, row, col)),
					Some(t) if t == '>' => Some(Lexeme::new(s_clone, LexemeType::ARROW, row, col)),
					Some(t) => {
						if t.is_numeric() {
							println!("Test = {:?}, c = {:?}", t, c );
							//negative numbers
							self.reader.pushback(c);
							self.reader.pushback(t);
							self.lex_number()
						}
						else {
							println!("Not numeric");
							self.reader.pushback(t);
							Some(Lexeme::new(s_clone, LexemeType::MINUS, row, col))
						}
					},
					None => Some(Lexeme::new(s_clone, LexemeType::MINUS, row, col))
				}
			},
			'/' => {
				let test = self.reader.read().unwrap();
				match test {
					'/' => {self.reader.skip_line(); self.lex()},
					_ => {
						self.reader.pushback(test); 
						Some(Lexeme::new(s_clone, LexemeType::DIVIDE, row, col))
					}
				}
			},
			'>' => {
				let test = self.reader.read().unwrap();
				match test {
					'=' => Some(Lexeme::new(s_clone, LexemeType::GTE, row, col)),
					_ => {
						self.reader.pushback(test); 
						Some(Lexeme::new(s_clone, LexemeType::GT, row, col))
					}
				}
			},
			'<' => {
				let test = self.reader.read().unwrap();
				match test {
					'=' => Some(Lexeme::new(s_clone, LexemeType::LTE, row, col)),
					_ => {
						self.reader.pushback(test); 
						Some(Lexeme::new(s_clone, LexemeType::LT, row, col))
					}
				}
			},
			'&' => {
				let test = self.reader.read().unwrap();
				match test {
					'&' => Some(Lexeme::new(s_clone, LexemeType::AND, row, col)),
					_ => {
						self.reader.pushback(test);
						self.reader.pushback(c);
						None
					}
				}
			},
			'|' => {
				let test = self.reader.read().unwrap();
				match test {
					'|' => Some(Lexeme::new(s_clone, LexemeType::OR, row, col)),
					_ => {
						self.reader.pushback(test);
						self.reader.pushback(c);
						None
					}
				}
			},
			'=' => {
				let test = self.reader.read().unwrap();
				match test {
					'=' => Some(Lexeme::new(s_clone, LexemeType::EQUALITY, row, col)),
					_ => {
						self.reader.pushback(test);
						Some(Lexeme::new(s_clone, LexemeType::ASSIGN, row, col))
					}
				}
			},
			'!' => {
				let test = self.reader.read().unwrap();
				match test {
					'=' => Some(Lexeme::new(s_clone, LexemeType::NOT_EQUALS, row, col)),
					_ => {
						self.reader.pushback(test);
						Some(Lexeme::new(s_clone, LexemeType::NOT, row, col))
					}
				}
			}
			_ => {
				self.reader.pushback(c);
				None
			}
		}
	}

	fn lex_number(&mut self) -> Option<Lexeme<S>>{
		println!("Lexing a number");
		//Get our character ready
		let row = self.reader.get_line();
		let col = self.reader.get_col(); 
		let c = self.reader.read().unwrap();
		println!("C is {:?}", c);

		let s_clone = self.source.clone();

		let mut string_buffer = String::new();

		if c != '-' && !c.is_numeric() {
			println!("Not a number");
			self.reader.pushback(c);
			return None
		}
		else{
			string_buffer.push(c);
		}
		
		loop{
			let character = self.reader.read();
			if let Some(c) = character{
				println!("loop c is {}", c );
				if !c.is_numeric() { 
					self.reader.pushback(c);
					break;
				}
				else{ string_buffer.push(c); }
			}
			else{
				break;
			}
		}
		println!("{:?}", string_buffer);
		println!("Returning Some(..) value from lex number");
		Some(Lexeme::new(s_clone, LexemeType::INTEGER(string_buffer.parse::<i32>().unwrap()), row, col))
	}

	fn lex_string(&mut self) -> Option<Lexeme<S>>{
		println!("Lexing string");
		//Get our character ready
		let row = self.reader.get_line();
		let col = self.reader.get_col(); 
		let c = self.reader.read().unwrap();

		let s_clone = self.source.clone();

		let mut string_buffer = String::new();
		// Only try to lex if we actually have a quoted string
		if c == '"' {
			loop{
				let character = self.reader.read();
				if let Some(c) = character{
					if c == '"'{
						break;
					}
					else if c == '\\'{
						string_buffer.push(self.reader.read().unwrap());
					}
					else{
						string_buffer.push(c);
					}
				}
				else{
					panic!("Unexpected end of input! Started String but never finished.");
				}
			}
			Some(Lexeme::new(s_clone, LexemeType::STRING(string_buffer), row, col))
		}
		else {
			self.reader.pushback(c);
			None
		}
	}

	fn lex_keywords_and_identifiers(&mut self) -> Option<Lexeme<S>>{
		//Get our character ready
		let row = self.reader.get_line();
		let col = self.reader.get_col(); 
		let s_clone = self.source.clone();
		
		let mut string_buffer = String::new();
		println!("In keyword lexer");

		let mut value = self.reader.read();
		while let Some(c) = value{
			println!("Read {}", c);
			if c.is_digit(36) || c.is_alphabetic() || c as u32 > 65536 || c == '-' || c == '_' {
				println!("Appending! {}", c);
				string_buffer.push(c);
				value = self.reader.read();
			}
			else{
				println!("Done building string");
				self.reader.pushback(c);
				value = None;
			}
		}

		println!("in lexer {:?}", string_buffer);

		match string_buffer.as_ref() {
			"var" => Some(Lexeme::new(s_clone, LexemeType::VAR, row, col)),
			"function" => Some(Lexeme::new(s_clone, LexemeType::FUNCTION, row, col)),
			"if" => Some(Lexeme::new(s_clone, LexemeType::IF, row, col)),
			"else" => Some(Lexeme::new(s_clone, LexemeType::ELSE, row, col)),
			"for" => Some(Lexeme::new(s_clone, LexemeType::FOR, row, col)),
			"while" => Some(Lexeme::new(s_clone, LexemeType::WHILE, row, col)),
			"do" => Some(Lexeme::new(s_clone, LexemeType::DO, row, col)),
			"true" => Some(Lexeme::new(s_clone, LexemeType::TRUE, row, col)),
			"false" => Some(Lexeme::new(s_clone, LexemeType::FALSE, row, col)),
			"null" => Some(Lexeme::new(s_clone, LexemeType::NULL, row, col)),
			"lambda" => Some(Lexeme::new(s_clone, LexemeType::LAMBDA, row, col)),
			"class" => Some(Lexeme::new(s_clone, LexemeType::CLASS, row, col)),
			"is" => Some(Lexeme::new(s_clone, LexemeType::IS, row, col)),
			_ => {
				Some(Lexeme::new(s_clone, LexemeType::IDENTIFIER(string_buffer), row, col))
			}
		}
	}
}

#[cfg(test)]
mod test {
	use super::*;
	use std::io::Cursor;

	#[test]
	fn single_lex_test(){
		let mut lexer = Lexer::new(Cursor::new("var"));
		assert_eq!(lexer.lex().unwrap().get_type(), LexemeType::VAR);
	}

	#[test]
	fn emoji_identifier_test(){
		let mut lexer = Lexer::new(Cursor::new("🚀"));
		assert_eq!(lexer.lex().unwrap().get_type(), LexemeType::IDENTIFIER("🚀".to_owned()));
	}

	#[test]
	fn lex_string_test(){
		let mut lexer = Lexer::new(Cursor::new("\"test string\""));
		assert_eq!(lexer.lex().unwrap().get_type(), LexemeType::STRING("test string".to_owned()));
	}

	#[test]
	fn lex_integer_test(){
		let mut lexer = Lexer::new(Cursor::new("5"));
		assert_eq!(lexer.lex().unwrap().get_type(), LexemeType::INTEGER(5));
	}

	#[test]
	fn lex_multi_integer_test(){
		let mut lexer = Lexer::new(Cursor::new("55"));
		assert_eq!(lexer.lex().unwrap().get_type(), LexemeType::INTEGER(55));
	}

	#[test]
	fn lex_negative_integer_test(){
		let mut lexer = Lexer::new(Cursor::new("-5"));
		assert_eq!(lexer.lex().unwrap().get_type(), LexemeType::INTEGER(-5));
	}

	#[test]
	fn lex_boolean_test(){
		let mut lexer = Lexer::new(Cursor::new("true false"));
		assert_eq!(lexer.lex().unwrap().get_type(), LexemeType::TRUE);
		assert_eq!(lexer.lex().unwrap().get_type(), LexemeType::FALSE);
	}

	#[test]
	fn lex_math_symbols_test(){
		let mut lexer = Lexer::new(Cursor::new("* / + - %"));
		assert_eq!(lexer.lex().unwrap().get_type(), LexemeType::MULTIPLY);
		assert_eq!(lexer.lex().unwrap().get_type(), LexemeType::DIVIDE);
		assert_eq!(lexer.lex().unwrap().get_type(), LexemeType::PLUS);
		assert_eq!(lexer.lex().unwrap().get_type(), LexemeType::MINUS);
		assert_eq!(lexer.lex().unwrap().get_type(), LexemeType::MODULO);
	}

	#[test]
	fn lex_negation_test(){
		let mut lexer = Lexer::new(Cursor::new("!true"));
		assert_eq!(lexer.lex().unwrap().get_type(), LexemeType::NOT);
		assert_eq!(lexer.lex().unwrap().get_type(), LexemeType::TRUE);
	}

	#[test]
	fn lex_boolean_ops_test(){
		let mut lexer = Lexer::new(Cursor::new("&& ||"));
		assert_eq!(lexer.lex().unwrap().get_type(), LexemeType::AND);
		assert_eq!(lexer.lex().unwrap().get_type(), LexemeType::OR);
	}

	#[test]
	fn lex_assign_equality_test(){
		let mut lexer = Lexer::new(Cursor::new("a==b a=b"));
		lexer.lex();
		assert_eq!(lexer.lex().unwrap().get_type(), LexemeType::EQUALITY);
		lexer.lex();
		lexer.lex();
		assert_eq!(lexer.lex().unwrap().get_type(), LexemeType::ASSIGN);
	}

	#[test]
	fn lex_negative_multi_integer_test(){
		let mut lexer = Lexer::new(Cursor::new("-55"));
		assert_eq!(lexer.lex().unwrap().get_type(), LexemeType::INTEGER(-55));
	}

	#[test]
	fn lex_line_col_check_test(){
		let mut lexer = Lexer::new(Cursor::new("true false"));
		if let Some(t) = lexer.lex(){
			assert_eq!(t.get_col(), 1);
			assert_eq!(t.get_line(), 1);
		}
		if let Some(f) = lexer.lex(){
			assert_eq!(f.get_col(), 6);
			assert_eq!(f.get_line(), 1);
		}
	}

	#[test]
	fn var_assignment_test(){
		let mut lexer = Lexer::new(Cursor::new("var a = b"));
		assert_eq!(lexer.lex().unwrap().get_type(), LexemeType::VAR);
		assert_eq!(lexer.lex().unwrap().get_type(), LexemeType::IDENTIFIER("a".to_owned()));
		assert_eq!(lexer.lex().unwrap().get_type(), LexemeType::ASSIGN);
		assert_eq!(lexer.lex().unwrap().get_type(), LexemeType::IDENTIFIER("b".to_owned()));
	}
}
